### DISCUSSION TITLE
Rebased commits by the web interface aren't signed

### DISCUSSION THREAD

**Discussion by BasixKOR** [community/community#11639] [Pull Requests]
Hi!

First of all, I'm more than happy to rebase my branch on GitHub without switching, pulling and all the jazz. However, when I rebase my branch on the PR by clicking 'Update with rebase', my signed commits become unsigned, resulting in an 'Unverified' badge if the user is using the [vigilant mode](https://docs.github.com/github/authenticating-to-github/displaying-verification-statuses-for-all-of-your-commits).

IIRC GitHub signs the commit that has been created on the website by GitHub's key, and I was wondering if we could do the same for rebasing too.

**Comment by BasixKOR** [comment]
This seemingly also applies to the "Merge with Rebase" feature too.

**Reply by redspiderweb1** [reply]
When using the web interface for version control, you might encounter an issue where rebased commits aren't signed, as noted in issue #11639. This can be a concern for projects requiring commit verification for security and integrity purposes. Unsigned commits can create challenges in maintaining a secure and trustworthy codebase.

**To address this, you can take several steps:**

**Manual Signing:** After rebase operations via the web interface, manually sign the commits using your local Git environment. This ensures that all commits, including those rebased, are properly signed.

**Automated Scripts:** Implement automated scripts or hooks in your development workflow to check for unsigned commits and prompt for manual signing before merging into the main branch.

**Configure Git Settings:** Ensure your Git settings are configured to sign commits by default. This can be set up in your global Git configuration or specific repository settings.

**Use of CI/CD Pipelines:** Integrate continuous integration/continuous deployment (CI/CD) pipelines to enforce commit signing policies. These pipelines can automatically verify the presence of commit signatures before allowing merges.

For businesses offering services like "[brochure design Dubai](https://www.redspider.ae/service/brochure-design-company-in-dubai/)," maintaining the integrity and security of your project repository is crucial. Ensuring all commits are signed, including those rebased via the web interface, helps protect your project from unauthorized changes and enhances the overall trustworthiness of your development process. Implementing these practices not only secures your codebase but also builds confidence with clients seeking reliable and professional design services.

**Reply by ebndev** [reply]
Hi @ashrafkhanak1, thanks for being a part of the GitHub Community!

Please note: we’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. Thank you for helping us maintain an authentic and beneficial space for everyone.

**Comment by dvakatsiienko** [comment]
This actually causes a lot of conflicts each time after local rebase once you pull **main** that is contains «rewritten without a sign» commits by Github, and rebase it onto existing branch that contains those commits in signed form. Thus, making Github Rebase feature pretty unusable which is bad.

**Reply by export-mike** [reply]
any updates on allowing github to sign your commits on a rebase?

**Reply by dvakatsiienko** [reply]
doesn't seem so :(

**Reply by devnoname120** [reply]
This will probably get deprecated as an “outdated issue” like all those long-running bugs below. 😆😆
https://github.com/github/roadmap/discussions/1014

@ankneis quickly locked the discussion of this announcement, probably in an effort to cens… uh I mean in order to listen more to their users, embody transparency, and give customers the chance to contribute their experience to help influence the roadmap.

**Comment by kshehadeh** [comment]
For what it's worth, this also seems to be an issue if you're using the API to rebase a PR which contains only signed commits.

**Comment by oliverbevan** [comment]
any update on this?

**Reply by mrsimonemms** [reply]
FWIW, it seems like a recent change. I've used this workflow for _AGES_ and it's only in the past week or two where I've seen unverified changes after merging the PR via the website

**Reply by Murali-Shl** [reply]
Hi, 
I am getting similar issue of getting the error 'branch is out-of-date with the base branch' and when I click the Rebase button in the UI, it is re-running the workflows (push commit and create PR actions).. 

Or I can click on bypass the checks to merge into master.

**Comment by REBELinBLUE** [comment]
Also seems to happen when you revert a PR, I am sure it used to work in the past

**Comment by alexandrmazur96** [comment]
Is any fix existing for this?

**Comment by danielnorberg** [comment]
Would love to see this fixed. It's pretty jarring to see a lot of `Unverified` commits in PRs, especially when not knowing it's due to the web UI rebase.

**Comment by zliang-akamai** [comment]
Still can observe the issue. Can you guys try to fix it?

**Comment by jorgediaz-lr** [comment]
I have also detected this issue.

**Comment by magneland** [comment]
Same.
Until the web interface is fixed, my workaround is to re-sign the commits locally:

```bash
gpg --list-secret-keys --keyid-format=long
git config --global user.signingkey <your key>
git rebase --exec 'git commit --amend --no-edit -n -S' -i <SHA before the one to fix>
```

**Reply by learosema** [reply]
I've put an alias for the rebase in my .gitconfig :)
```
[alias]
    resign = "!f() { git rebase --exec 'git commit --amend --no-edit -n -S' -i \"$1\"; }; f"
```

unintended pun

**Reply by i-nadia** [reply]
For those who already rebased through the web interface :
```shell
git pull
git commit --amend --no-edit --signoff
git push --force-with-lease
```

**Reply by relay-arik** [reply]
> I've put an alias for the rebase in my .gitconfig :)
> 
> ```
> [alias]
>     resign = "!f() { git rebase --exec 'git commit --amend --no-edit -n -S' -i \"$1\"; }; f"
> ```
> 
> unintended pun


Can also do, if you want to also verify the commit you pass to it, the only edge case is if $1 is the first commit

```
[alias]
    resign = "!f() { git rebase --exec 'git commit --amend --no-edit -n -S' -i \"$1^\"; }; f"
```

**Reply by mrsimonemms** [reply]
Quitting your job seems a bit of an overreaction 😄

**Comment by dannyrife** [comment]
Hit this again today.  Used the script above to fix.  Any update?

**Comment by microhobby** [comment]
Yeah, still broken. Some updated? It's annoying to have to do it manually when using GitHub.

**Comment by maxim** [comment]
Yeah, the pull request merge buttons no longer seem to sign my commits. It used to work a few months ago, and now it doesn't. I'm using an SSH Signing Key.

**Comment by azizur** [comment]
Any update on this? I just spotted the same issue.

**Comment by pektezol** [comment]
@magneland 's command works well, but I really don't want to force-push everytime I have to rebase. Why would the commits be unverified if it's on Github's own page and pull request, AND the commits are already signed..

Can we get some update on this?

**Reply by EduardoRT** [reply]
It's also impossible to do so with protected branches that require a pull-request + blocking force pushes for the base branch.

**Comment by karrakoliko** [comment]
Any update on this?

**Comment by tiriana** [comment]
Does anyone have a workaround for this? Right now I resign commits locally and push --force-with-lease them

**Comment by gorbak25** [comment]
+1

**Comment by ForsakenRei** [comment]
I hit this issue as well. And based on [Github doc](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#:~:text=Signature%20verification%20for%20rebase,pull%20request%27s%20base%20branch), since Github has no access to the private key then it cannot sign the commits, you can only merge it locally and then push as a workaround.

But as above I remembered it worked before, now I'm wondering how it works before...

**Reply by bfren** [reply]
> Github has no access to the private key then it cannot sign the commits

I don't fully understand this part of that document, as GitHub clearly does have access to the private key - when I create a normal commit through the web interface it *is* marked as verified. 

I get that merge commits are slightly different, but they are still created by user action, ie clicking a button in the same way as a standard commit.

**Reply by ForsakenRei** [reply]
Well me either. Currently I have to resign the commits after PR merged or I just merge locally if there is no PR required, this is definitely not ideal...

**Reply by ctsstc** [reply]
Yeah I was going to say I believe if you do a merge it signs it properly, but a rebase does not, so it seems to have its own key that since you're the user signed in it signs it with that.

**Reply by beatrixada** [reply]
Correct, commits made via the web UI are signed with GitHub's key. This is what I would assume they'd do for rebases in the web UI, but I guess maybe they feel uncomfortable signing commits that might be made by multiple users with the same key?

**Reply by bfren** [reply]
(or even users that have made unsigned commits I guess?)

**Reply by ForsakenRei** [reply]
I cannot say for others but I sign all my own commits.

**Reply by bfren** [reply]
I feel on repos where multiple people contribute, it would be pretty easy to say 'if all the commits in this PR are signed, then sign the merge commit by the user doing the merge'.

If all the commits in a PR were *not* signed however that wouldn't be possible.

**Reply by matfax** [reply]
I think the reason for this change is that GitHub now also discards the signatures of rebase merges. In theory, this doesn't require any change to the commit, so the signature could remain unchanged. But for some reason, GitHub adds the merging party into the author list - without review, without adopted code suggestions, just for clicking the (auto) merge button. This is nonsensical. The commit only has to change with squashes, hence requiring a new signature. I'm also not sure what happened to the "partially verified" feature. It just shows up as unverified now. GitHub could solve this by re-signing the commit with a key to which you grant access for web signing. Perhaps, also add passkey-encryption or access. Or use the smartcard feature that most browsers have and let users store their keys on their personal hardware.

**Reply by jarkad** [reply]
> The commit only has to change with squashes, hence requiring a new signature.

Rebasing also changes the commit's identity, which requires re-signing. Rebase works by creating new commits with identical diffs and messages, but different parents; old commits are garbage-collected.

**Comment by jramosf** [comment]
I've raised a support ticket to see if we can get an official response from GitHub on this one.

**Reply by jramosf** [reply]
I am not sure if I can paste the verbatim contents of the response from the support engineer here, but basically I got response and they have open issues on Engineering but no specific commitment at the moment.

Pasting my own response as I've already closed the ticket.

![image](https://github.com/community/community/assets/9470856/bc64ca1d-f7de-4cf0-8ed9-ad7655eec3c0)

I'll keep an eye on this thread and in GitHub changelog in case it gets implemented in the future.

**Reply by rj-au** [reply]
@jramosf did you get around to logging a ticket? If so, any updates?

**Comment by bhatiashivam** [comment]
Is anyone aware of any official response from Github?  Right now I resign commits locally and push with `--force-with-lease`.

**Reply by bfren** [reply]
Only what @jramosf posted a few days ago - obviously they're aware of it but have no plans to do anything about it.

**Comment by dvins** [comment]
Maddening, open two years. Instead of doing simple updates in UI from a phone or tablet have to lug around my laptop and fire up full Git to be able to merge a stack of PRs.

**Reply by tiriana** [reply]
We disabled rebase because of this.
Atlassian has tickets open since 2014. 2 years is like a warmup.

**Comment by dvins** [comment]
@ebndev Is someone on the GitHub team responsible for this issue? Its nigh on two and a half years and frankly its an infosec security and provenance matter.

**Reply by dvins** [reply]
@ebndev Can we please get a little attention on this from you even or another GitHub community manager, if only to point us in the direction of whom else we should be trying to speak with? As stated previously, this is an infosec security and provenance matter. This thread, with more than 300 bumps and thumbs ups, has now been open without feedback from GitHub for more than two and a half years.

**Reply by ebndev** [reply]
Hi @dvins, I've brought this to the attention of the Community Manager of this category so they can assess the feedback and forward it to the proper GitHub team.

**Here's what you can expect moving forward ⏩**
- Your input will be carefully reviewed and cataloged by members of our product teams. 
  - Due to the high volume of submissions, we may not always be able to provide individual responses.
  - Rest assured, your feedback will help chart our course for product improvements.
- GitHub staff may reach out for further clarification or insight. 
  - We may 'Answer' your discussion if there is a current solution, workaround, or roadmap/changelog post related to the feedback.

We appreciate you bringing attention to an older however, for future incidents-- please know that we do actively check the "Product Feedback" label and follow the processes mentioned above. Community members can upvote, comment, and react with discussions that resonate with them to voice their support on feature requests.

While we can't promise that every suggestion will be implemented, we want to emphasize that your feedback is instrumental in guiding our decisions and priorities. Thanks for understanding! ⭐

**Reply by dvins** [reply]
Thank you @ebndev, much appreciated. Do be aware that there is no tag on this that indicates this is an infosec and security related issue though. As others have indicated, this bug allows others to work around protection rules and results in undue provenance when performing merges.

**Reply by dvins** [reply]
@ebndev Whom / how do we get this issue elevated? It's been four plus years now, this is a highly important especially with stacked pull request support coming soon. Likewise, this issue still does not have any security or infosec tags associated with it, so it continues to fly under the radar. Last, this is truly an enterprise class issue, and still remains unresolved. Starting to make me wonder if GitLab is more responsive for such matters?

**Reply by matfax** [reply]
This is the one issue that just makes me lose the most faith in GitHub. It's obviously a security concern. What surprises me about it is that no single enterprise partner of GitHub has poked them right away when it emerged. It's not a direct vulnerability, but this leads to sloppy workarounds and cleanup attempts that then pose a threat and fall under GitHub's radar, or it leads to to ignorance of the whole feature that in theory solves important questions of trust in security modeling; something regularly done by enterprise-scale customers of GitHub. GitHub allows policy enforcement based on a feature that works in unexpected ways.

**Reply by EduardoRT** [reply]
@matfax This is my personal account, but I have access to an Enterprise subscription and support, I've poked them about this and many other issues, they just don't care about it, GitHub support, even at the enterprise level is terrible.

**Reply by dvins** [reply]
@ebndev Whom / how do we get this issue elevated? It's been four and half plus years now, this is a highly important especially with stacked pull request support coming soon. Likewise, this issue still does not have any security or infosec tags associated with it, so it continues to fly under the radar. Last, this is truly an enterprise class issue, and still remains unresolved.

**Comment by matfax** [comment]
If you're merging a PR as bypass listee, this even ignores the "Require signed commits" protection rule without giving notice.

**Comment by BoscoDomingo** [comment]
October 2024, this is still an issue. 2 years and almost 8 months...

**Comment by ferrata** [comment]
Hey GitHub team! I just saw an announcement that [Persistent commit signature verification now in public preview](https://github.blog/changelog/2024-11-13-persistent-commit-signature-verification-now-in-public-preview/). Does it mean that this issue is going to be addressed?

**Reply by kingg22** [reply]
I still have this problem, only the commits with push to master have verified

![image](https://github.com/user-attachments/assets/3c1815c6-5f7c-4db0-8f79-287dea1fc9b6)
![image](https://github.com/user-attachments/assets/2a649590-5f27-47b9-9b6d-9f33f95e3135)

**Comment by devnoname120** [comment]
Not even `gh pr merge 123` works. This command generates the merge/rebased/squash commit from GitHub's side rather than locally, so it has the same problems as with the web interface.

**Comment by AndreasPresthammer** [comment]
My team has enabled required signed commits for our `main` branch. We've hit this stumbling block where we no longer can rebase PRs into the main branch due to this exact issue. We then had to fall back to using squash commits or merge commits into the main branch.

I find it very odd that github has no problem signing squash commits and merge commits with their own key, but doesn't support re-signing rebase commits into main. Having a rule in the github rebase logic to only allow resigning of verified commits sounds fairly trivial to do. This would allow us to start using the rebase-to-main flow again and still retain the requirement for signed commits in `main`.

Alternatively github could avoid re-creating the commits, but instead allow for fast forward merge to main which would retain the PRs unchanged, already signed, commits to go into main.

The situation as it stand reduces the quality of our main branch, since in my teams workflow we often want to retain the commits from the PR into main. Sadly we cannot do this today with signed commits requirement in main with this current limitation.

We _can_ retain history using merge commits, but this causes a noisy `main` history which would prefer not to have.

**Reply by vRune4** [reply]
> Alternatively github could avoid re-creating the commits, but instead allow for fast forward merge to main which would retain the PRs unchanged, already signed, commits to go into main.

For me that would be the best solution. (I don't see why this isn't the default to begin with)

**Reply by YoshiRulz** [reply]
Something like #6414?

**Comment by diegodorado** [comment]
any updates on this?

**Reply by bfren** [reply]
As if!

**Comment by PrayagS** [comment]
3+ years and not a single concrete update 😔

**Comment by renanborgez** [comment]
this is very annoying, so much time and this is still an issue

**Comment by tylerhunt** [comment]
I've taken to manually merging PRs from the command line. When doing a `git merge --ff-only <branch>` and `git push`, GitHub will detect that the branch has been merged and close the PR. However, you do need to take care to ensure that whatever is on the branch has been pushed to GitHub before doing the merge or the PR will not be closed, and you'll be left with a mess to clean up (having to close the PR without showing it as having been merged).

**Reply by thunze** [reply]
Thanks for the pointer!

I was curious and ran some tests, and it seems that neither non-ff merges nor unpushed additional commits on the branch actually prevent GitHub from correctly detecting the merge and closing the associated PR. For example, I merged https://github.com/thunze/playground/pull/5 from the command line, despite merging an [additional unpushed commit](https://github.com/thunze/playground/commit/66503b6848a14d0aa0094deb0e68ff0abd965acc) and resolving a merge conflict while merging.

Where things _do_ fall apart, i.e., PRs don't get closed, is when you don't merge all commits that have already been pushed to the remote branch associated with the PR. This can for example happen when you rebase locally before merging locally or when you [squash-merge](https://github.com/thunze/playground/pull/8) locally.

There are even more examples here: https://github.com/thunze/playground/pulls?q=.

**Comment by rubberduck203** [comment]
Just chiming in here to say that we also expected the rebase button in the UI to resign with Github's key the way it does when you perform a squash merge.

**Comment by thoran** [comment]
It's coming up to 4 years since this was opened. 👍

**Comment by njelich** [comment]
Bump, please!

**Comment by mu88** [comment]
Does anyone have a good workaround in the meantime, while the underlying problem has not been resolved? I have a branch protection rule that forces the signing of commits – I now have to manually deactivate this every time a PR is created, merge the PR, and then reactivate it. Is it possible to create any exception in the branch protection rules for PRs?

**Reply by ParadoxV5** [reply]
Push directly to ~~prod~~ _the base_ is technically an option, but it’s not like one can give external contributors direct push access.

The workaround, then, is to merge by fast-forwarding: repoint the base branch to the PR’s tip using local Git, which should not recreate the contributed commits without signatures.
If the PR is not up to date, the contributor will have to rebase locally (which recreates their commits with fresh signatures) and re-(force-)push first.

Interestingly, the GitHub interface doesn’t have fast-forward-only merges; it can only require PRs to be up to date before merging.

**Reply by mu88** [reply]
Okay, but that's nothing someone can do without a Git installation - I was think about a pure web-based solution. Here's my concrete problem:
- I use [Renovate's branch auto-merge strategy](https://docs.renovatebot.com/key-concepts/automerge/#branch-vs-pr-automerging) to keep my dependencies up to date.
- Renovate creates a new commit which triggers all necessary builds to run. These builds are also configured as status checks within the branch protection rule.
- Because I renamed a build step (which is marked as mandatory in my branch protection rule), the required build was not found by Renovate. Therefore it filed a PR. This PR shows the violated status check due to the missing build.
- I fixed the branch protection rule to use the correct build name.
- The PR is now ready to merge because all status checks are green (due to the fixed naming).
- When trying to merge the PR, I cannot merge it due to the missing Git signature of the rebased commit - so I have to disable the branch protection rule, merge the PR, and enable it again.

**Comment by SeanCassiere** [comment]
So, going into 2026 and this still has no fix. It would be nice if GitHub focused a bit more on these Git-related issues.

**Reply by Oxygen66** [reply]
> So, going into 2026 and this still has no fix. It would be nice if GitHub focused a bit more on these Git-related issues.

Yaaa ... 4 Years without any update on this.

**Reply by BoscoDomingo** [reply]
Blasphemy! How are they not going to focus on AI™ and pushing all these features users **clearly and desperately** need!1!11!!!

**Comment by azizur** [comment]
Come on GitHub, you can vibe code a solution using CoPilot.

Imagine the headline CoPilot can fix old bugs without an engineer!

**Comment by cacarico** [comment]
This isn't really a bug, it's a technical limitation. When you rebase locally, your Git client re-applies and signs each commit using your private key, which never leaves your machine. When GitHub performs the rebase server-side, it rewrites the commits (new SHAs, new parent references) but has no access to your private key, so it can't sign them...

On repos that requires verified commits I always disable rebase and squash and merge

**Reply by thunze** [reply]
No, GitHub could sign with its own key. That's e.g. also what happens with merge commits created when merging a PR via the GitHub UI.

**Reply by portante** [reply]
I agree.  This makes little sense.

Why can't GitHub create a private key for my account, which I agree to add the public key like any other private key I add (which I can revoke at any time) and have the rebase and merge work seemlessly with signing?

**Reply by randompixel** [reply]
It can also sign things auto-magically when inside a Codespace for you as long as you have some settings on. I was hoping the same would apply here.

**Comment by mustafaabobakr** [comment]
# any updates here

**Reply by mustafaabobakr** [reply]
<img src="https://static.klipy.com/ii/d6b0ce929193df3c242ac34b5654d2ce/bb/6d/1gEuUnQH.gif" />

**Comment by willjones9** [comment]
I found this issue today when trying out stacked PRs. The idea of stacked PRs is great, but if using them breaks commit verification, it's a no from us

**Reply by dvins** [reply]
This issue has been open for nearly five years now, so have lost any hope this, the similar issue with stacked PRs you point out, along with this additional, attribution issue, **Stacked PR squash merges can combine the PR author's name with the merger's email** https://github.com/github/gh-stack/issues/371 will ever get fixed.

**Comment by salehablrmubeen-source** [comment]
This would definitely be a nice improvement. When GitHub rebases a branch through the web interface, it rewrites the commits, so the original GPG/SSH signatures can't be preserved. If GitHub could automatically re-sign the rewritten commits with its verified web-flow key (similar to other commits created through the UI), it would help maintain trust and avoid the "Unverified" badge for users with vigilant mode enabled.

Until then, the only reliable workaround is to perform the rebase locally and re-sign the rewritten commits before pushing.

**Reply by EduardoRT** [reply]
That's the thing, some people excuse this as they "don't have your private key so they can't sign for you" but they do it all the time for web commits so they clearly have a flow for this, it should be recognized as a different commit state that it was signed using GitHub's web key so that at least it doesn't break workflows.

**Reply by soroushm** [reply]
The only way I can think of to solve this is with a GitHub Action that re-signs the rebase. However, that means a GPG key would have to be included in the action, which creates a window of exposure and a potential security risk.

**Reply by matfax** [reply]
> The only way I can think of to solve this is with a GitHub Action that re-signs the rebase. However, that means a GPG key would have to be included in the action, which creates a window of exposure and a potential security risk.

I think a GitHub App could sign it with a service/bot account. GitHub Actions can't do it, as you wrote, without some user's private key, which destroys trust in the signing process itself and defies it's purpose.

**Reply by nyoma-diamond** [reply]
@EduardoRT 
> but they do it all the time for web commits so they clearly have a flow for this, it should be recognized as a different commit state that it was signed using GitHub's web key so that at least it doesn't break workflows.

I think a potential concern is if the rebase adjusts commits by _multiple authors_. Logically if all commits were made by the same user as the one performing the rebase then resigning them is trivial as you suggest. 

**_However_** if there were _other_ contributors to the sequence prior to rebase (or for example a maintainer rebases another contributor's PR branch to bring it up to date, where said maintainer has no commits on that branch), then GitHub re-signing the commits introduces a major concern wherein the act of user A---who didn't make all (or potentially any) of the commits---performing a rebase causes contributions by user B---who did make commits---to be re-signed **without actually getting their approval.** I.e., user A makes a change and then unilaterally re-signs user B's work. This is a HUGE risk. This may also be confounded further once co-authorship is taken into consideration (though co-authors usually don't sign commits; only the primary authors do. Multi-author signing is its own can of worms). This is not to mention the fact that it's also possible to edit commits mid-rebase, and therefore possible to modify another user's work in the process (though iirc GitHub doesn't provide any way to do so via web interface. But still _technically_ possible). Either way, the risk of re-signing an unrelated user's work is definitely a no-no which I assume GitHub wouldn't want to touch with a ten foot pole.

That said, GitHub _could_ reasonably split the difference by only re-signing the commits of the user performing the rebase. Not ideal, but still potentially an improvement.

**Reply by soroushm** [reply]
@nyoma-diamond & @matfax since rebasing changes the commit hash, the signature/trust of the original commit no longer applies the rebased commit is technically a different commit.

Co-authorship doesn't help in this case either, since the commit itself has changed. The co-author would need to sign the new commit again, and at that point it wouldn't be the same commit you originally authored.

**Reply by matfax** [reply]
Different issues that are being discussed, namely squashing and rebasing.

First squashing: If you use a bot for signing, the bot has to establish trust based on its own model. You could make all user approvals part of the approval model, even create a branched co-signing system on a separate platform. But the overhead of making this separate platform/GitHub App secure, to establish it as a reliable and trustable extension is probably too big if you don't have some corporation backing/funding it from the start. Similar issue with what the GitHub Actions have become: large maintainers prefer official and in-house only, because that can be trusted.

For squashing, it's far easier for GitHub themselves to just establish a new web-based signing system where every single co-author needs to co-sign. Integrity of trust would then have to be maintained by preserving the rebased branches if squashing was used.

Squashing, however, can already be disabled on a repo level, therefore not be excluded from the trust chain as a consideration. I think the biggest user frustration is that rebasing alone, without any squashing, already leads to the signature issue and also that branch protection rules can be bypassed with then destroyed signatures. If that bypass was impossible, at least users would first have to resign the rebase, manually rebase locally. Local rebasing automatically re-signs, it's just GitHub's web rebase that ignores it. You can create web-based changes that are auto-signed, but web-based rebases are not.

There's one limitation where the current web-based system reaches a limitation even if it applied to rebases: it only works if the rebasing user is the only commit author in the branch. Most PRs are single author. Those that aren't would require the same expansion like the squash co-authoring, so that a rebase becomes a process where users receive a notification to resign, so they all have to visit the PR and approve, which then accesses their private key to resign their own commits in particular. The same would happen when maintainers try to rebase contributor PRs.

A possible extension of this could be that GitHub securely stores the private key for backend-based automatic signing of rebases and squashes. It doesn't violate trust since rebasing and squashing are automatic, predictable processes that don't really require user review. The rebase process can automatically note the original ref in the commit messages.

This means there are just two major gaps in GitHub's signature system. One is the lack of web-based single author rebase (not squash) signing, the other is the rules that can be bypassed (at least last I've checked). Only after that, the co-authoring of squashes would have to be solved. And merge commits always remain a third, unlikable, option.

**Reply by nyoma-diamond** [reply]
@soroushm 
> since rebasing changes the commit hash, the signature/trust of the original commit no longer applies the rebased commit is technically a different commit.

Maybe I'm missing something, but personally I don't see how changing a commit hash differs practically from creating an entirely brand new commit via web commit. In both situations we're using the GitHub web interface to synthesize new commits. We already rely on GitHub to maintain the trust chain when making modifications through the web interface (e.g., web commits, merges, etc.), so at least in the single-contributor case I don't see how this precludes the ability for GitHub to re-sign rebases. It's not meaningfully different
